guide_drafts/trait_derives.md

```diff
@@ -0,0 +1,927 @@
+}
+```
+
+If we try to insert records without deriving `Insertable`, we would get the following error.
+
+```rust
+error[E0277]: the trait bound `&diesel_demo::models::NewUser<'_>: diesel::Insertable<diesel_demo::schema::users::table>` is not satisfied
+  --> src/bin/main.rs:29:10
+   |
+29 |         .values(&new_user)
+   |          ^^^^^^ the trait `diesel::Insertable<diesel_demo::schema::users::table>` is not implemented for `&diesel_demo::models::NewUser<'_>`
+```
+
+Now **this** is a great error message.
+The compiler tells us outright that `Insertable` isn't implemented for our `NewUser` struct.
+
+## Identifiable
+
+Certain database operations,
+such as table [associations](#associations)  and updates, 
+require that rows be uniquely identifiable.
+Implementing the `Identifiable` trait on your struct will define which columns (primary keys)
+make your struct uniquely identifiable.
+To Implement `Identifiable`, 
+annotate your struct with [`#[derive(Identifiable)]`][identifiable_doc] 
+
+[identifiable_doc]: https://docs.diesel.rs/diesel/associations/trait.Identifiable.html
+
+By default, `Identifiable` will assume the primary key is a column named `id`.
+If your table's primary key is named differently,
+you can annotate the table with the attribute `#[primary_key(some_field_name)` or `#[primary_key(some_field_name, another_field_name)`.
+The `Identifiable` trait will assume that the annotated struct will be named in the singular form of the table it corresponds to.
+If the name differs you may use the `#[table_name="some_table_name"]` attribute annotation.
+Having the `Identifiable` trait provides us the `id()` method on our models,
+which returns the value of our record's primary key.
+
+In the following example,
+we will look at some of the behavior `Identifiable` provides for us.
+We will add the annotation to our `User` struct.
+We will then attempt to get the value of the first record's primary key by calling `id()` and also update the first and last name of our user.
+
+### Example
+
+```rust
+// File: src/models.rs
+
+use schema::users;
+
+#[derive(Identifiable, Queryable)]
+pub struct User {
+    pub id: i32,
+    pub first_name: String,
+    pub last_name: String,
+    pub email: String,
+}
+
+#[derive(Insertable)]
+#[table_name="users"]
+pub struct NewUser<'a> {
+    pub first_name: &'a str,
+    pub last_name: &'a str,
+    pub email: &'a str,
+ }
+```
+
+```rust
+// File: src/main.rs
+
+use diesel::prelude::*;
+
+fn main() {
+  let new_user = NewUser { 
+      first_name: "Gordon", 
+      last_name: "Freeman", 
+      email: "gordon.freeman@blackmesa.co" 
+  };
+
+  diesel::insert_into(users::table)
+      .values(&new_user)
+      .execute(&db_connection)
+      .expect("Error inserting row into database");
+
+  let all_users = users.load::<User>(&db_connection)
+      .expect("Error loading users");
+
+  println!("User count: {}", all_users.len()); // > User count: 1
```

Comment:
I'd put the return value in a comment on the _next_ line. Some manuals use `//=> foo` to say "foo" was returned FYI

Reply:
In these cases you could also use `assert_eq!` (like `assert_eq!(all_users.len(), 1)`), though